runners/apex/src/main/java/org/apache/beam/runners/apex/ApexYarnLauncher.java

```diff
@@ -0,0 +1,379 @@
+ * <p>
+ * The proxy takes the DAG and communicates with the Hadoop services to launch
+ * it on the cluster.
+ */
+public class ApexYarnLauncher {
+  private static final Logger LOG = LoggerFactory.getLogger(ApexYarnLauncher.class);
+
+  public AppHandle launchApp(StreamingApplication app) throws IOException {
+
+    List<File> jarsToShip = getYarnDeployDependencies();
+    StringBuilder classpath = new StringBuilder();
+    for (File path : jarsToShip) {
+      if (path.isDirectory()) {
+        File tmpJar = File.createTempFile("beam-runners-apex-", ".jar");
+        createJar(path, tmpJar);
+        tmpJar.deleteOnExit();
+        path = tmpJar;
+      }
+      if (classpath.length() != 0) {
+        classpath.append(':');
+      }
+      classpath.append(path.getAbsolutePath());
+    }
+
+    EmbeddedAppLauncher<?> embeddedLauncher = Launcher.getLauncher(LaunchMode.EMBEDDED);
+    DAG dag = embeddedLauncher.getDAG();
+    app.populateDAG(dag, new Configuration(false));
+
+    Attribute.AttributeMap launchAttributes = new Attribute.AttributeMap.DefaultAttributeMap();
+    launchAttributes.put(YarnAppLauncher.LIB_JARS, classpath.toString().replace(':', ','));
+    LaunchParams lp = new LaunchParams(dag, launchAttributes);
+    lp.cmd = "hadoop " + ApexYarnLauncher.class.getName();
+    HashMap<String, String> env = new HashMap<>();
+    env.put("HADOOP_USER_CLASSPATH_FIRST", "1");
+    env.put("HADOOP_CLASSPATH", classpath.toString());
+    lp.env = env;
+    return launchApp(lp);
+  }
+
+  protected AppHandle launchApp(LaunchParams params) throws IOException {
+    File tmpFile = File.createTempFile("beam-runner-apex", "params");
+    tmpFile.deleteOnExit();
+    FileOutputStream fos = new FileOutputStream(tmpFile);
+    SerializationUtils.serialize(params, fos);
+    fos.close();
+    if (params.getCmd() == null) {
+      ApexYarnLauncher.main(new String[] {tmpFile.getAbsolutePath()});
+    } else {
+      String cmd = params.getCmd() + " " + tmpFile.getAbsolutePath();
+      ByteArrayOutputStream consoleOutput = new ByteArrayOutputStream();
+      LOG.info("Executing: {} with {}", cmd, params.getEnv());
+
+      ProcessBuilder pb = new ProcessBuilder("bash", "-c", cmd);
+      Map<String, String> env = pb.environment();
+      env.putAll(params.getEnv());
+      Process p = pb.start();
+      ProcessWatcher pw = new ProcessWatcher(p);
+      InputStream output = p.getInputStream();
+      InputStream error = p.getErrorStream();
+      while (!pw.isFinished()) {
+        IOUtils.copy(output, consoleOutput);
+        IOUtils.copy(error, consoleOutput);
+      }
+      if (pw.rc != 0) {
+        throw new RuntimeException("Failed to run: " + cmd + " (exit code " + pw.rc + ")" + "\n"
+            + consoleOutput.toString());
+      }
+    }
+    return new AppHandle() {
+      @Override
+      public boolean isFinished() {
+        // TODO: interaction with child process
```

Comment:
Just checking that these `TODO` comments are intended for fixing in later PRs, right, not blocking this one?